src/collab.ts

```diff
@@ -108,64 +188,294 @@ export function receiveUpdates(
   })
 }
 
+// Receive remote commits, generating a ChangeSet to represent the difference
+// between the last remote committed value and the last confirmed value or
+// doing nothing if the values match.
+export function receiveCommits(state: EditorState, values: string[]) {
+  const collabState = state.field(collabField)
+  if (!values.length) return state.update({annotations: [collabReceive.of(collabState)]})
+  const update = diff(collabState.confirmedValue, state.toText(values[values.length - 1]))
+  if (!update) return state.update({annotations: [collabReceive.of(collabState)]})
+  return receiveUpdates(state, [update], 0)
+}
+
+// Create a commit, marking a local save point whose value can't be re-written.
+// This allows for special handling of remote updates received before the
+// commit has been confirmed by the authority, giving us an opportunity to
+// preserve their changes so they aren't completely overwritten by our commits.
+export function createCommit(state: EditorState, value: string) {
+  const collabState = state.field(collabField)
+
+  if (value !== state.doc.toString())
+    throw new Error("Committed value does not match editor state.")
+
+  // Determine how many updates should exist in the new commit, excluding
+  // unconfirmedUpdates that have already been allocated to previous commits.
+  const committedUpdateCount = sum(collabState.unconfirmedCommits)
+  const uncommittedUpdateCount = collabState.unconfirmedUpdates.length - committedUpdateCount
+
+  // Return a transaction updating collab state.
+  return state.update({
+    annotations: [
+      collabReceive.of(
+        new CollabState(
+          collabState.version,
+          collabState.unconfirmedUpdates,
+          // Add to unconfirmedCommits.
+          [...collabState.unconfirmedCommits, uncommittedUpdateCount],
+          collabState.confirmedValue,
+          // Set confirmedToUnconfirmed if not already set.
+          collabState.confirmedToUnconfirmed || ChangeSet.empty(collabState.confirmedValue.length)
+        )
+      )
+    ]
+  })
+}
+
+// Confirm any number of previously marked commits. This updates collab state
+// based on the count associated with the confirmed commits.
+export function confirmCommits(state: EditorState, values: string[]) {
+  const collabState = state.field(collabField)
+  const count = values.length;
+
+  // Cannot confirm more commits than those in the current list.
+  if (count > collabState.unconfirmedCommits.length)
+    throw new Error("Count greater than unconfirmed commits.")
+
+  // If no commits are being confirmed, return old collab state.
+  if (!count) return state.update({annotations: [collabReceive.of(collabState)]})
+
+  // Determine which and how many updates are being confirmed.
+  const confirmedCommits = collabState.unconfirmedCommits.slice(0, count)
+
+  // Begin constructing new collab state.
+  let version = collabState.version
+  const unconfirmedUpdates = collabState.unconfirmedUpdates.slice()
```

Comment:
```suggestion
  const unconfirmedUpdates = [...collabState.unconfirmedUpdates];
```

I assume we're doing this cause we want to clone the array, correct?